### PR TITLE
CorrelationAlert model added

### PR DESCRIPTION
**Pull request by riysaxen-amzn**
### Description
* Creates a model class for GenericAlert i.e BaseAlert class which can be extended later on by monitor based alerts
* Creates a model class for CorrelationAlert which extends BaseAlert class
* Tests added to support CorrelationAlerts
 
### Issues Resolved
[Alerts in Correlations](https://github.com/opensearch-project/security-analytics/issues/988)
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+import org.opensearch.core.xcontent.XContentBuilder
+import org.opensearch.core.xcontent.XContentParser
+import org.opensearch.core.xcontent.XContentParserUtils
+import java.io.IOException
+import java.time.Instant
+
+/** CorrelationAlert and Alert can extend the UnifiedAlert class to inherit the common fields and behavior
+ * of UnifiedAlert class.
+ */
+open class UnifiedAlert(
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt:22]: should this be called BaseAlert?

File: src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    override fun writeTo(out: StreamOutput) {
+        super.writeTo(out)
+        out.writeStringCollection(correlatedFindingIds)
+        out.writeString(correlationRuleId)
+        out.writeString(correlationRuleName)
+    }
+    override fun asTemplateArg(): Map<String, Any?> {
+        val superTemplateArgs = super.asTemplateArg()
+        val correlationSpecificArgs = mapOf(
+            "correlatedFindingIds" to correlatedFindingIds,
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt:86]: nit: use Constant variables defined below

File: src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt (hunk 1 above)
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt:22]: why is Alert class not extending this class?

File: src/test/kotlin/org/opensearch/commons/alerting/CorrelationAlertTests.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import org.opensearch.commons.alerting.model.Alert
+import org.opensearch.commons.alerting.model.CorrelationAlert
+import org.opensearch.commons.utils.getJsonString
+import org.opensearch.core.common.bytes.BytesArray
+import org.opensearch.core.common.bytes.BytesReference
+import org.opensearch.core.common.io.stream.InputStreamStreamInput
+import org.opensearch.core.xcontent.NamedXContentRegistry
+import java.time.temporal.ChronoUnit
+
+class CorrelationAlertTests {
```
Comment (eirsep) [inline src/test/kotlin/org/opensearch/commons/alerting/CorrelationAlertTests.kt:17]: plz add test for serialization roundtrip..

**Review by eirsep — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt (hunk 1 above)
Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt:22]: we can do that at a later stage, when we will start Push Notifications and  whenever we onboard MonitorBased Alerts onto it. Right now focus for this PR is just Correlation based alerts

File: src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt (hunk 2 above)
Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt]: updated

Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt]: updated

File: src/test/kotlin/org/opensearch/commons/alerting/CorrelationAlertTests.kt (hunk 3 above)
Comment (riysaxen-amzn) [reply src/test/kotlin/org/opensearch/commons/alerting/CorrelationAlertTests.kt:18]: updated

File: src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt (hunk 1 above)
Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/UnifiedAlert.kt]: changed to BaseAlert

File: src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    open val endTime: Instant? = null,
+    open val acknowledgedTime: Instant? = null,
+    open val errorMessage: String? = null,
+    open val severity: String,
+    open val actionExecutionResults: List<ActionExecutionResult>
+) : Writeable, ToXContent {
+
+    init {
+        if (errorMessage != null) {
+            require((state == Alert.State.DELETED) || (state == Alert.State.ERROR) || (state == Alert.State.AUDIT)) {
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt:41]: Why would alert state Deleted mandate a non-null error message

File: src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                        while (xcp.nextToken() != XContentParser.Token.END_ARRAY) {
+                            correlatedFindingIds.add(xcp.text())
+                        }
+                    }
+                    CORRELATION_RULE_ID -> correlationRuleId = xcp.text()
+                    CORRELATION_RULE_NAME -> correlationRuleName = xcp.text()
+                }
+            }
+
+            val unifiedAlert = parse(xcp, version)
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt:121]: Which parse() method is called? Recursive?

Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/CorrelationAlert.kt:121]: yes, so in this case, the parse function of CorrelationAlert is utilizing the parse function of its superclass BaseAlert to handle the parsing of fields common to both types of alerts. Then, it proceeds to parse the additional fields specific to CorrelationAlert

File: src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt (hunk 4 above)
Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt:41]: yeah this one's in Alert state class also, so eventually Monitor based Alert (i.e Alert class) will also extend this BaseClass, therefore.

File: src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                "Attempt to create an alert with an error in state: $state"
+            }
+        }
+    }
+
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
```
Comment (sbcd90) [inline src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt:48]: are there writeable tests for these constructors? like this. https://github.com/opensearch-project/common-utils/blob/main/src/test/kotlin/org/opensearch/commons/alerting/model/WriteableTests.kt

File: src/test/kotlin/org/opensearch/commons/utils/TestHelpers.kt [hunk 7]
```diff
@@ -16,7 +17,11 @@ import java.io.ByteArrayOutputStream
 fun getJsonString(xContent: ToXContent): String {
     ByteArrayOutputStream().use { byteArrayOutputStream ->
         val builder = XContentFactory.jsonBuilder(byteArrayOutputStream)
-        xContent.toXContent(builder, ToXContent.EMPTY_PARAMS)
+        if (xContent is CorrelationAlert) {
```
Comment (sbcd90) [inline src/test/kotlin/org/opensearch/commons/utils/TestHelpers.kt:20]: do we need this change? because this is not part of alerting package.

Comment (riysaxen-amzn) [reply src/test/kotlin/org/opensearch/commons/utils/TestHelpers.kt:20]: yeah I'm using `getJsonString` method to convert correlation alert Json into string on line 87 in src/test/kotlin/org/opensearch/commons/alerting/CorrelationAlertTests.kt therefore changed this to make it work for CorrelationAlert

File: src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt (hunk 6 above)
Comment (riysaxen-amzn) [reply src/main/kotlin/org/opensearch/commons/alerting/model/BaseAlert.kt:48]: no, but here src/test/kotlin/org/opensearch/commons/alerting/CorrelationAlertTests.kt in `test correlation parse function` we are validating such scenarios

**Review by eirsep — APPROVED** [review]

**Review by sbcd90 — APPROVED** [review]

**Pull request merged**